PlutoHelperAgent/SharedFunctions.m

```diff
@@ -0,0 +1,236 @@
+                                                             
+                                                             );
+    
+    if (pwAccessStatus == errSecSuccess) {
+        
+        NSData* data = [NSData dataWithBytes:pwData length:pwLength];
+        
+        NSString* password = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+        
+        SecKeychainAttribute attrs2[] = {
+            
+            { kSecAccountItemAttr, 0, NULL }
+            
+        };
+        
+        SecKeychainAttributeList attributes2 = { 1, attrs2 };
+        
+        OSStatus unAccessStatus = SecKeychainItemCopyContent(itemRef, NULL, &attributes2, NULL, NULL);
+        
+        if (unAccessStatus == errSecSuccess) {
+            
+            NSLog(@"Username retrived from Apple Keychain");
+            
+            NSData* data8 = [NSData dataWithBytes:attributes2.attr->data length:attributes2.attr->length];
+            
+            NSString* usernamedata = [[NSString alloc] initWithData:data8 encoding:NSUTF8StringEncoding];
+            
+            NSArray *returnStrings = [NSArray arrayWithObjects:
+                                      usernamedata,
+                                      password,
+                                      nil];
+            
+            return returnStrings;
+            
+            
+        } else {
+            
+            NSLog(@"Username not retrived from Apple Keychain");
+        }
+        
+    } else {
+        
+        NSLog(@"Keychain read failed: %@", SecCopyErrorMessageString(pwAccessStatus, NULL));
+        
+    }
+    
+    if (pwData) SecKeychainItemFreeContent(NULL, pwData);  // Free memory
+    
+    NSArray *failedReturnStrings = [NSArray arrayWithObjects:
+                                    @"Username not loaded",
+                                    @"Password not loaded",
+                                    nil];
+    
+    return failedReturnStrings;
+    
+}
+
++ (NSString *) login_to_project_server {
+    
+    NSArray *dataFromKeychain = [self load_data_from_keychain];
+    
+    NSString *URLToUse = [NSString stringWithFormat: @"%@/api/login", [[NSUserDefaults standardUserDefaults] stringForKey:@"project_locker_url"]];
+    
+    NSURL *url = [NSURL URLWithString:URLToUse];
+    NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
+    NSURLSession *session = [NSURLSession sessionWithConfiguration:config];
+    
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:url];
+    request.HTTPMethod = @"POST";
+    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
+    NSLog(@"Request %@", request);
+    
+    NSDictionary *dictionary = @{@"username": dataFromKeychain[0], @"password": dataFromKeychain[1]};
+    NSError *error = nil;
+    NSData *data = [NSJSONSerialization dataWithJSONObject:dictionary
```

Comment:
I would suggest factoring out the serialize-to-JSON and upload-to-server code into their own module as we'll need that a lot for this project

Reply:
Done